File: 백준/Gold/2098. 외판원 순회/외판원 순회.py
from sys import stdin, maxsize
from collections import defaultdict
input = stdin.readline

def dfs(cur,vis):
    if vis == (1 << n) - 1: # 1000(2) - 1 -> 0111(2) 0, 1, 2 방문 확인 
        if cost[cur][0]:
            return cost[cur][0]
        else:
            return maxsize
    if (cur,vis) in dp:
        return dp[(cur,vis)]
    min_c = maxsize
    for nxt in range(1,n):
        if not cost[cur][nxt] or vis & (1 << nxt): # 방문했거나 값이 없거나
            continue
        c = dfs(nxt, vis | (1 << nxt)) + cost[cur][nxt]
        min_c = min(min_c,c)
    dp[(cur,vis)] = min_c
    return min_c

n = int(input())
cost = [list(map(int,input().split())) for _ in range(n)]
dp = defaultdict(int)
print(dfs(0,1))

File: 백준/Gold/2098. 외판원 순회/외판원 순회.java
import java.io.*;
import java.util.*;

public class Main {
	static int N, cost[][], dp[][], INF = 10000000;

	public static void main(String[] args) throws Exception {
		//System.setIn(new FileInputStream("res/input.txt"));
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st;
		N = Integer.parseInt(br.readLine());
		cost = new int[N][N];
		for (int i = 0; i < N; i++) {
			st = new StringTokenizer(br.readLine());
			for (int j = 0; j < N; j++) {
				cost[i][j] = Integer.parseInt(st.nextToken());
			}
		}
		dp = new int[N][1 << N];
		System.out.println(TSP(0, 1 << 0));
	}

	static int TSP(int cur, int flag) {
		if (flag == (1 << N) - 1) {
			if (cost[cur][0] != 0)
				return cost[cur][0];
			else
				return INF;
		}
		if (dp[cur][flag] != 0)
			return dp[cur][flag];
		int minDist = INF;
		for (int nxt = 1; nxt < N; nxt++) {
			if (cost[cur][nxt] == 0)
				continue;
			if ((flag & (1 << nxt)) == 0) { 
				minDist = Math.min(minDist, cost[cur][nxt] + TSP(nxt, flag | (1 << nxt)));
			}
		}
		dp[cur][flag] = minDist;
		return minDist;
	}
}